Local open in types

johnyob opened the pull request:
## Motivation
**Related issue**: #9484

Currently the AST permits `Foo.(...)` in patterns and expressions but not types. There are cases where such syntax could reduce the verbosity of types without requiring an `open` in the module. 

## Description

The PR extends `Parsetree.core_type` and `Typedtree.core_type` with a `(P|T)typ_open` constructor.
The parser is extended to permit these opens in any delimited types (parenthesised, constructor and class arguments, variants, objects, extensions).

```ocaml
(* Some examples *)
module F = struct
  type t = int
  type s = string
  type r = unit
end

type paren = F.(t * s * r)

type ('a, 'b, 'c) triple = 'a * 'b * 'c

type args = F.(t, s, r) triple

type obj = F.< x : t; y : s; z : r >

type var = F.[ `X of t | `Y of s | `Z of r ]
```

The outcome tree encoding of the type (e.g. `paren`) is:
```
# type paren = F.(t * s * r);;
type paren = F.t * F.s * F.r
```
Personally I'm not entirely happy with this result (the ideal output would be `F.(t * s * r)`), however, I'm not sure how to achieve this (so any pointers would be greatly appreciated!)

johnyob committed:
Implement local open for types

gasche commented on parsing/parser.mly:
```diff
@@ -3431,6 +3441,27 @@ atomic_type:
+      { mk_ptyp_class ~loc:$sloc tys cid }
+  | mktyp( /* begin mktyp group */
+      mkrhs(mod_ext_longident) DOT delimited_type 
+        { Ptyp_open ($1, $3) }
```
Nitpick: I like menhir's ability to name things instead of using $3 to mean something, and I would encourage you to use it for new parsing rules. (Parts of the current parser were inherited from the previous ocamlyacc grammar which did not support this.)

johnyob replied:
Will do -- I also prefer this style

johnyob replied:
[fd02ca9](https://github.com/ocaml/ocaml/pull/12044/commits/fd02ca917db07e4aa19de0175bcf4f285e2e52ec)

gasche commented on typing/tast_iterator.ml:
```diff
@@ -548,6 +548,7 @@ let typ sub {ctyp_loc; ctyp_desc; ctyp_env; ctyp_attributes; _} =
   | Ttyp_variant (list, _, _) -> List.iter (sub.row_field sub) list
   | Ttyp_poly (_, ct) -> sub.typ sub ct
   | Ttyp_package pack -> sub.package_type sub pack
+  | Ttyp_open (_, _, t) -> sub.typ sub t
```
We probably want some `iter_loc` call in this case as well.

johnyob replied:
[393c0e7](https://github.com/ocaml/ocaml/pull/12044/commits/393c0e7171baf3045f3725c3fd230e4276012bf3)

gasche commented on typing/tast_mapper.ml:
```diff
@@ -748,6 +748,8 @@ let typ sub x =
     | Ttyp_package pack ->
         Ttyp_package (sub.package_type sub pack)
+    | Ttyp_open (path, mod_ident, t) ->
+        Ttyp_open (path, mod_ident, sub.typ sub t)
```
`map_loc sub mod_ident`?

johnyob replied:
2ebf83c

gasche commented on typing/typetexp.ml:
```diff
@@ -380,6 +380,46 @@ let transl_type_param env styp =
+  vs
+
+
+type policy = Fixed | Extensible | Univars
```
The stuff above looks like an erroneous resolution of a merge conflict with the recent TyVarEnv refactoring. i would remove all of it.

johnyob replied:
Yes, this was fixed in commit 2bc28b40852a173e57e08a567e9646d77d5bd4e9. I can fix it, but I'm more than happy for the commits to be squashed when merging

gasche reviewed (commented):
The implementation looks very reasonable to me, what I would expect. I don't think that there is a simple way to magically guess `open` on the way back when printing types, and I don't think it should be a requirement for this feature.

gasche commented:
I see a potential issue with the `A.< (object type) >` form, which is a lexer-level conflict with BER MetaOCaml's `.< (code) >.` syntax. Could we leave out the object-type open form for now, out of abundance of caution?

(If we believe that this creates a jarring inconsistency with the open form for bracketed polymorphic variants, then I think we could also leave those out. Almost all the value comes from the parenthesized form anyway.)

Relatedly to syntax bikeshedding: should we consider having the `let Path in <type>` form as well? I think that a reasonable answer is "no" (as in the current status of the PR), because (1) we also don't support this for patterns, and (2) there are no keyword constructions in the language of type expressions right now, so it would feel a bit out of place.

nojb commented:
I don't want to be a wet cloth (and I am not strongly opposed to this PR), but going over the discussion in #9484, I confess I don't find the justification of the feature very convincing: this feature "could drastically reduce the verbosity of complex type signatures". The real-life example given in that PR is:
```ocaml
type ('a, 'b, 'c, 'r_type, 'w_type) pipe =
  ('a, 'b, [> Stream.Readable.kind] as 'r_ype) Stream.Readable.subtype ->
  (('b, 'c, [> Stream.Writable.kind] as 'w_type) Stream.Writable.subtype as 'w_stream) ->
  'w_stream
```
that becomes
```ocaml
type ('a, 'b, 'c, 'r_type, 'w_type) pipe = Stream.(
  Readable.(('a, 'b, [> kind] as 'r_type) subtype) ->
  Writable.(('b, 'c, [> kind] as 'w_type) subtype) as 'w_stream ->
  'w_stream
)
```
which does not seem to make such a big difference to me. **And**, local opens of types have the same downside as local open of values: they harm readability because you don't know if the names that appear inside the local open are shadowed or not.

Having said that, one may argue this feature on other grounds (eg regularity of the language), but I thought I would chime in with my 2c: personally, I don't see myself using this feature very often, if at all.

gasche commented:
I think that the feature does make the language more regular (we get asked about it from users expecting it to just work, and the UX is better when people are not randomly surprised by stuff not working), and has similar tradeoffs to `open` constructors for other grammatical categories (structures, expressions, patterns). It is also fairly simple to implement, see the PR which is mostly boilerplate with zero new logic in the type-system.

nojb commented:
> I think that the feature does make the language more regular (...) It is also fairly simple to implement, see the PR which is mostly boilerplate with zero new logic in the type-system.

Agreed, this is good enough for me!

nojb commented on typing/tast_iterator.ml:
```diff
@@ -548,6 +548,9 @@ let typ sub {ctyp_loc; ctyp_desc; ctyp_env; ctyp_attributes; _} =
   | Ttyp_poly (_, ct) -> sub.typ sub ct
   | Ttyp_package pack -> sub.package_type sub pack
+  | Ttyp_open (_, mod_ident, t) -> 
+    iter_loc sub mod_ident;
+    sub.typ sub t
```
Nitpick: two space indent to match the other cases in this function. Thanks!

nojb commented on typing/typetexp.ml:
```diff
@@ -651,6 +659,12 @@ and transl_type_aux env policy styp =
             pack_txt = p;
            }) ty
+  | Ptyp_open (mod_ident, t) -> 
+    let path, new_env = 
+      !type_open Asttypes.Fresh env loc mod_ident
+    in
+    let cty = transl_type new_env policy t in
+    ctyp (Ttyp_open (path, mod_ident, cty)) cty.ctyp_type
```
Idem.

nojb reviewed (commented):
I went over the code, it looks quite correct. I don't have an opinion about object types, but I thought that the direction for the MetaOCaml patch was to use attribute syntax instead? (Perhaps I misremember the discussion.)

lpw25 commented:
I like the general feature, but this one just seems confusing:
```ocaml
type args = F.(t, s, r) triple
```
Could we perhaps get rid of the constructor case?

Octachron commented on typing/typetexp.mli:
```diff
@@ -15,6 +15,7 @@
 
 (* Typechecking of type expressions for the core language *)
 
+open Asttypes
```
Personal taste: I would rather avoid adding a global open and use `Asttypes.override_flag` below.

johnyob commented:
> I see a potential issue with the `A.< (object type) >` form, which is a lexer-level conflict with BER MetaOCaml's `.< (code) >.` syntax. Could we leave out the object-type open form for now, out of abundance of caution?
> 
> (If we believe that this creates a jarring inconsistency with the open form for bracketed polymorphic variants, then I think we could also leave those out. Almost all the value comes from the parenthesized form anyway.)
> 
> Relatedly to syntax bikeshedding: should we consider having the `let Path in <type>` form as well? I think that a reasonable answer is "no" (as in the current status of the PR), because (1) we also don't support this for patterns, and (2) there are no keyword constructions in the language of type expressions right now, so it would feel a bit out of place.

I concur with this caution here, will remove 👍

johnyob commented:
> I like the general feature, but this one just seems confusing:
> 
> ```ocaml
> type args = F.(t, s, r) triple
> ```
> 
> Could we perhaps get rid of the constructor case?

Personally I'm a fan of this syntax since I'd prefer it over `F.((t, s, r) triple)`, but to avoid bike-shedding over syntax I can remove it in this PR

goldfirere commented:
Just coming across this and wanted to say that I've reached for this syntax before and it will be nice to have it. Thanks, @johnyob!

gasche commented:
@johnyob I missed the fact that you had updated the PR by removing the more controversial examples of the syntax. Great, thanks! I need to do another review round and I hope that this would be good to merge.

gasche commented on parsing/parser.mly:
```diff
@@ -3392,44 +3392,75 @@ tuple_type:
+    - variant types               [ `A ]
+    - extension                   [%foo ...]
+*)
+delimited_type_without_object_type:
```
A short comment to explain why you are single-casing object types in the definition here could help. (The difference of course is that they do not support `Foo.` qualification.)

gasche replied:
Note: I am okay with the current state, but I would actually slightly prefer to:

1. Call this one `delimited_type_supporting_local_open`
2. Moving the `Ptyp_extension` out of this group, so that we stop supporting `M.[%foo]` as a local open. I see no strong need for that syntax, and I can see hypothetical situations where we would want to define a different meaning for quantified expansions. (I'm not sure, but we regularly get bitten by having accepted too many things that we now want to give a different meaning to, so I would not support a syntax unless there is a clear need.)

gasche commented on parsing/parser.mly:
```diff
@@ -3392,44 +3392,75 @@ tuple_type:
+  | tys = actual_type_parameters
+    HASH
+    cid = mkrhs(clty_longident)
+      { mktyp ~loc:$sloc (Ptyp_class (cid, tys)) }
```
Why are the two cases above moved out of the common `mktyp( ... )` wrapper below? (My guess would be that they were different when you wanted to allow local-open on parameter tuples, but that you reverted this aspect and could in fact make the diff smaller and the code slightly more factorized by putting them back in the wrapper.)

gasche reviewed (commented):
I did have another look and I do have some comments again. Overall this is looking nice and, if/when @johnyob is available to do a rebase and answer the last comment round, I think we should move toward merging.

( @johnyob if you don't have time to work on this anymore, let us know and I could probably push a rebase myself. )

johnyob commented:
> ( @johnyob if you don't have time to work on this anymore, let us know and I could probably push a rebase myself. )

I'm happy to work on addressing your comments and rebasing this weekend :)

johnyob committed:
Bootstrap parser

johnyob committed:
Pleasing the linter

johnyob committed:
Update dependencies

johnyob commented:
@gasche I've rebased, hopefully this should be in good shape for merging :)

gasche commented:
@johnyob could you add a Changes entry in the "Language features" section? Ideally having a short example of the new syntax would be nice.

gasche commented:
(I was planning to merge and do this myself but I see that you are working on the CI results.)

johnyob committed:
Add changelog entry

johnyob commented:
`make alldepend` is a noop locally, but the CI is failing 🤔 Even running `tools/ci/actions/check-alldepend.sh` locally gives me a successful result

Octachron commented on parsing/ast_helper.ml:
```diff
@@ -114,6 +115,8 @@ module Typ = struct
         | Ptyp_package(longident,lst) ->
             Ptyp_package(longident,List.map (fun (n,typ) -> (n,loop typ) ) lst)
+        | Ptyp_open (mod_ident, core_type) ->
+            Ptyp_open (mod_ident, core_type)
```
This is missing a `loop core_type`. Otherwise
```ocaml
let f: type a. Int.( a -> a ) = fun x -> x;;
```
fails to transform `a` into `'a` yielding the following error:

> Error: Unbound type constructor a

Octachron commented:
It looks like the dependency graph varies between the ARM64 and AMD backend, and that's the part causing the failure. You can discard the dependency change in `asmcomp` to make the linter happy (while I open an issue on this matter).

gasche commented:
(I don't understand the CI situation regarding dependencies. Some explanations in an issue would help, or @Octachron is free to merge the current PR when he feels that the CI is in an acceptable state.)

johnyob commented:
> (I don't understand the CI situation regarding dependencies. Some explanations in an issue would help, or @Octachron is free to merge the current PR when he feels that the CI is in an acceptable state.)

I must also admit that I am similarly perplexed about the CI situation 😅

Octachron commented:
I will push the fix for the dependency update. This is unrelated to the CI, this is a matter of the dependency graph being different when building the compiler on a M1.

Octachron committed:
fix dependencies

Octachron committed:
fix bootstrapped parser synchronisation

Octachron commented:
I  have pushed the dependency fix, and resynced the bootstrapped parser. I propose to do a squash merge once the CI is green (to avoid keeping the linter-related commits).

gasche commented:
Thanks! I agree that squashing is fine.

johnyob commented:
Thanks for the help @Octachron! 🙏 Much appreciated :)

hyphenrf commented:
Hello, quick question regarding this PR.. Considering the following
```ocaml
type t = What of Foo.(a * b * c)
type u = Unboxed of Foo.a * Foo.b * Foo.c
       | Boxed of (Foo.a * Foo.b * Foo.c)
```
What is `What` equivalent to, `Unboxed` or `Boxed`? It seems to be the latter,
```ocaml
# module A = struct type t = int end
  type b = B of A.(t * t)
  ;;
module A : sig type t = int end
type b = B of (A.t * A.t)
```
and if that's the case... why? I'd be (often accidentally) paying runtime performance for syntactic convenience.. plus it's strictly less expressive like this.

lpw25 commented:
```ocaml
type t = What of Foo.(a * b * c)
```
If it can be done without butchering the parser, this should really just be an error.

gasche commented:
> If it can be done without butchering the parser, this should really just be an error.

cc @let-def / @jmid

let-def commented:
Tricky issue, the syntax of types in constructor arguments is very "irregular".
It's possible to achieve that by slightly refactoring the type syntax:

 - factor out `atomic_type_without_open` from `atomic_type`
 - change `constructor_arguments` to not allow open syntax at the top

With this we can reject the open syntax if there is only one argument:

```ocaml
# type t = What of Foo.(a * b * c);;
Error: Syntax error
# type t = What of Foo.(a * b * c) * foo;;
Error: Unbound module Foo
```

This will also reject the open syntax for non-tuples:

```ocaml
# type t = What of Foo.(t);;
Error: Syntax error
```

From there, LRgrep can easily produce a nicer error message. And before it is production-ready, we can add a separate rule for explaining why this syntax is rejected.

The parser is not butchered too much:
```diff
--- a/parsing/parser.mly
+++ b/parsing/parser.mly
@@ -3164,9 +3164,12 @@ generalized_constructor_arguments:
 ;

 constructor_arguments:
-  | tys = inline_separated_nonempty_llist(STAR, atomic_type)
+  | atomic_type_without_open
     %prec below_HASH
-      { Pcstr_tuple tys }
+      { Pcstr_tuple [$1] }
+  | ty = atomic_type STAR tys = inline_separated_nonempty_llist(STAR, atomic_type)
+    %prec below_HASH
+      { Pcstr_tuple (ty :: tys) }
   | LBRACE label_declarations RBRACE
       { Pcstr_record $2 }
 ;
@@ -3468,7 +3471,7 @@ delimited_type:
     { $1 }
 ;

-atomic_type:
+atomic_type_without_open:
   | type_ = delimited_type
       { type_ }
   | mktyp( /* begin mktyp group */
@@ -3479,10 +3482,6 @@ atomic_type:
       HASH
       cid = mkrhs(clty_longident)
         { Ptyp_class (cid, tys) }
-    | mod_ident = mkrhs(mod_ext_longident)
-      DOT
-      type_ = delimited_type_supporting_local_open
-        { Ptyp_open (mod_ident, type_) }
     | QUOTE ident = ident
         { Ptyp_var ident }
     | UNDERSCORE
@@ -3491,6 +3490,17 @@ atomic_type:
   { $1 } /* end mktyp group */
 ;

+%inline atomic_type:
+  | atomic_type_without_open
+  | mktyp(
+      mod_ident = mkrhs(mod_ext_longident)
+      DOT
+      type_ = delimited_type_supporting_local_open
+        { Ptyp_open (mod_ident, type_) }
+    )
+  { $1 }
+;
+
```

Alternatively, we could check for this construction in the semantic action and warn if necessary, but for this we would need to keep track of more details of the concrete syntax, like parentheses (that could be done with `reloc_typ` in https://github.com/johnyob/ocaml/blob/547539831025f56569b547a3b564cf106d8d321c/parsing/parser.mly#L3421).
This is probably better if we want to reject, or warn, only about `Foo.(a * b)` syntax.

gasche commented:
Weird, I am pretty sure that I tried exactly your proposal (or something very very close) and my attempt died by a thousand reduce/reduce conflicts. Thanks a lot! I will send is as a separate PR.
